File: CTNotificationContent/Views/CTCaptionedImageView.m

#import "CTCaptionedImageView.h"
#import "UIImage+CTImage.h"

static const float kCaptionHeight = 18.f;
static const float kSubCaptionHeight = 20.f;
static const float kSubCaptionTopPadding = 8.f;
static const float kBottomPadding = 18.f;
static const float kCaptionLeftPadding = 10.f;
static const float kCaptionTopPadding = 8.f;
static const float kImageBorderWidth = 1.f;
static const float kImageLayerBorderWidth = 0.4f;
static float captionHeight = 0.f;

@interface CTCaptionedImageView ()

@property (nonatomic, strong, nullable, readwrite) NSString *actionUrl;
@property (nonatomic, strong) NSString *caption;
@property (nonatomic, strong) NSString *subcaption;
@property (nonatomic, strong) NSString *imageUrl;

@property (nonatomic, strong) UIImageView *imageView;
@property (nonatomic, strong) UILabel *captionLabel;
@property (nonatomic, strong) UILabel *subcaptionLabel;

@end

@implementation CTCaptionedImageView

+ (CGFloat)captionHeight {
    if (captionHeight <= 0) {
        captionHeight = kCaptionHeight+kSubCaptionHeight+kBottomPadding;
    }
    return captionHeight;
}

- (instancetype _Nonnull)initWithFrame:(CGRect)frame
                               caption:(NSString * _Nullable)caption
                            subcaption:(NSString * _Nullable)subcaption
                              imageUrl:(NSString * _Nonnull)imageUrl
                             actionUrl:(NSString * _Nullable)actionUrl {
    
    self = [super initWithFrame:frame];
    if (self) {
        self.imageUrl = imageUrl;
        self.caption = caption;
        self.subcaption = subcaption;
        self.actionUrl = actionUrl;
        [self setup];
    }
    return self;
}

- (void)setup {
    CGFloat viewWidth = self.frame.size.width;
    CGFloat viewHeight = self.frame.size.height;
    CGSize imageViewSize = CGSizeMake(viewWidth, viewHeight-([[self class] captionHeight]));
    
    // gyrations to draw a corresponding gray border below the image
    self.imageView = [[UIImageView alloc] initWithFrame:CGRectMake(0.f-kImageBorderWidth, 0.f-kImageBorderWidth, imageViewSize.width + (kImageBorderWidth*2), imageViewSize.height)];
    self.imageView.contentMode = UIViewContentModeScaleAspectFit;
    self.imageView.backgroundColor = [UIColor colorWithRed:0.95 green:0.95 blue:0.95 alpha:1.0];
    self.imageView.layer.borderColor = [[UIColor lightGrayColor] CGColor];
    self.imageView.layer.borderWidth = kImageLayerBorderWidth;
    self.imageView.layer.masksToBounds = YES;
    [self addSubview:self.imageView];
    [self loadImage];
    
    self.captionLabel = [[UILabel alloc]initWithFrame:CGRectMake(kCaptionLeftPadding, kCaptionTopPadding + imageViewSize.height, viewWidth - kCaptionLeftPadding * 2, kCaptionHeight)];
    self.captionLabel.textAlignment = NSTextAlignmentLeft;
    self.captionLabel.adjustsFontSizeToFitWidth = NO;
    self.captionLabel.font = [UIFont boldSystemFontOfSize:16.f];
    self.captionLabel.textColor = [UIColor blackColor];
    self.captionLabel.text = self.caption;
    [self addSubview:self.captionLabel];
    
    self.subcaptionLabel = [[UILabel alloc]initWithFrame:CGRectMake(kCaptionLeftPadding, imageViewSize.height + kCaptionHeight + kSubCaptionTopPadding, viewWidth - kCaptionLeftPadding * 2, kSubCaptionHeight)];
    self.subcaptionLabel.textAlignment = NSTextAlignmentLeft;
    self.subcaptionLabel.adjustsFontSizeToFitWidth = NO;
    self.subcaptionLabel.font = [UIFont systemFontOfSize:12.f];
    self.subcaptionLabel.textColor = [UIColor lightGrayColor];
    self.subcaptionLabel.text = self.subcaption;
    [self addSubview:self.subcaptionLabel];
}

- (void)showPlaceholderImage {
    UIImage *placeholder = [UIImage ct_imageWithString:self.caption color:nil size:self.imageView.frame.size];
    self.imageView.image = placeholder;
}

- (void)loadImage {
    if (!self.imageUrl) return;
    
    NSURL *attachmentURL = [NSURL URLWithString:self.imageUrl];
    NSURLSession *session = [NSURLSession sessionWithConfiguration:[NSURLSessionConfiguration defaultSessionConfiguration]];
    [[session downloadTaskWithURL:attachmentURL
                completionHandler:^(NSURL *temporaryFileLocation, NSURLResponse *response, NSError *error) {
                    if (error != nil) {
                        dispatch_async(dispatch_get_main_queue(), ^{
                            [self showPlaceholderImage];
                        });
#ifdef DEBUG
                        NSLog(@"unable to download image: %@", error.localizedDescription);
#endif
                        
                    } else {
                        NSError *imageError;
                        
                        NSData *imageData = [NSData dataWithContentsOfURL:temporaryFileLocation
                                                                  options:kNilOptions
                                                                    error:&imageError];
                        
                        UIImage *image = [UIImage imageWithData:imageData];
                        
                        dispatch_async(dispatch_get_main_queue(), ^{
                            if (image) {
                                self.imageView.image = image;
                            } else {
                                [self showPlaceholderImage];
                            }
                            
                        });
#ifdef DEBUG
                        if (error != nil) {
                            NSLog(@"unable to download image: %@", error.localizedDescription);
                        }
#endif
                    }
                }] resume];
    
}

@end

File: CTNotificationContent/Views/CTCaptionedImageView.swift
import UIKit

struct CaptionedImageViewComponents {
    var caption: String = ""
    var subcaption: String = ""
    var imageUrl: String = ""
    var actionUrl: String = ""
    var bgColor: String = ""
    var captionColor: String = ""
    var subcaptionColor: String = ""
    var bgColorDark: String = ""
    var captionColorDark: String = ""
    var subcaptionColorDark: String = ""
}

class CTCaptionedImageView : UIView {
    var components = CaptionedImageViewComponents()

    private var imageView: UIImageView = {
        let imageView = UIImageView()
        imageView.contentMode = .scaleAspectFit
        imageView.layer.borderColor = UIColor.lightGray.cgColor
        imageView.layer.masksToBounds = true
        imageView.translatesAutoresizingMaskIntoConstraints = false
        return imageView
    }()
    private var captionLabel: UILabel = {
        let captionLabel = UILabel()
        captionLabel.textAlignment = .left
        captionLabel.adjustsFontSizeToFitWidth = false
        captionLabel.font = UIFont.boldSystemFont(ofSize: 16.0)
        captionLabel.translatesAutoresizingMaskIntoConstraints = false
        return captionLabel
    }()
    private var subcaptionLabel: UILabel = {
        let subcaptionLabel = UILabel()
        subcaptionLabel.textAlignment = .left
        subcaptionLabel.adjustsFontSizeToFitWidth = false
        subcaptionLabel.font = UIFont.systemFont(ofSize: 12.0)
        subcaptionLabel.translatesAutoresizingMaskIntoConstraints = false
        return subcaptionLabel
    }()

    init(components: CaptionedImageViewComponents) {
        super.init(frame: .zero)

        self.components = components
        setUpViews()
        setupConstraints()
    }
    
    override init(frame: CGRect) {
        super.init(frame: .zero)
    }
    
    required init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    func setUpViews() {
        addSubview(imageView)
        addSubview(captionLabel)
        addSubview(subcaptionLabel)

        CTUtiltiy.checkImageUrlValid(imageUrl: components.imageUrl) { [weak self] (imageData) in
            DispatchQueue.main.async {
                if imageData != nil {
                    self?.imageView.image = imageData
                    self?.activateImageViewContraints()
                }
            }
        }
        captionLabel.setHTMLText(components.caption)
        subcaptionLabel.setHTMLText(components.subcaption)
        updateInterfaceColors()
    }
    
    @objc public override func traitCollectionDidChange(_ previousTraitCollection: UITraitCollection?) {
            super.traitCollectionDidChange(previousTraitCollection)
            
            // Check if iOS 12+ API is available before using it
            if #available(iOS 12.0, *) {
                if traitCollection.userInterfaceStyle != previousTraitCollection?.userInterfaceStyle {
                    updateInterfaceColors()
                }
            }
    }
    
    func updateInterfaceColors() {
        // Check if device is in dark mode (iOS 12+)
        let isDarkMode: Bool
        
        if #available(iOS 12.0, *) {
            isDarkMode = traitCollection.userInterfaceStyle == .dark
        } else {
            // For iOS versions before 12.0,using light mode colors since dark mode wasn't officially supported
            isDarkMode = false
        }
        
        imageView.backgroundColor = UIColor(hex: isDarkMode ? components.bgColorDark : components.bgColor)
        captionLabel.textColor = UIColor(hex: isDarkMode ? components.captionColorDark : components.captionColor)
        subcaptionLabel.textColor = UIColor(hex: isDarkMode ? components.subcaptionColorDark : components.subcaptionColor)
    }
    
    func setupConstraints() {
        NSLayoutConstraint.activate([
            captionLabel.topAnchor.constraint(equalTo: bottomAnchor, constant: -(CTUtiltiy.getCaptionHeight() - Constraints.kCaptionTopPadding)),
               captionLabel.leadingAnchor.constraint(equalTo: leadingAnchor, constant: Constraints.kCaptionLeftPadding),
               captionLabel.trailingAnchor.constraint(equalTo: trailingAnchor, constant: -Constraints.kCaptionLeftPadding),
               captionLabel.heightAnchor.constraint(equalToConstant: Constraints.kCaptionHeight),
               
               subcaptionLabel.topAnchor.constraint(equalTo: bottomAnchor, constant: -(Constraints.kSubCaptionHeight + Constraints.kSubCaptionTopPadding)),
               subcaptionLabel.leadingAnchor.constraint(equalTo: leadingAnchor, constant: Constraints.kCaptionLeftPadding),
               subcaptionLabel.trailingAnchor.constraint(equalTo: trailingAnchor, constant: -Constraints.kCaptionLeftPadding),
               subcaptionLabel.bottomAnchor.constraint(equalTo: bottomAnchor, constant: -Constraints.kSubCaptionTopPadding),
               subcaptionLabel.heightAnchor.constraint(equalToConstant: Constraints.kSubCaptionHeight)
           ])
    }
    
    func activateImageViewContraints() {
        NSLayoutConstraint.activate([
           imageView.topAnchor.constraint(equalTo: topAnchor, constant: -Constraints.kImageBorderWidth),
           imageView.leadingAnchor.constraint(equalTo: leadingAnchor, constant: -Constraints.kImageBorderWidth),
           imageView.trailingAnchor.constraint(equalTo: trailingAnchor, constant: Constraints.kImageBorderWidth),
           imageView.bottomAnchor.constraint(equalTo: bottomAnchor, constant: -CTUtiltiy.getCaptionHeight())
        ])
    }
}

extension UIColor {
    public convenience init?(hex: String) {
        let red, green, blue: CGFloat

        if hex.hasPrefix("#") {
            let start = hex.index(hex.startIndex, offsetBy: 1)
            let hexColor = String(hex[start...])
            if hexColor.count == 6 {
                let scanner = Scanner(string: hexColor)
                var hexNumber: UInt64 = 0

                if scanner.scanHexInt64(&hexNumber) {
                    red = CGFloat((hexNumber & 0xff0000) >> 16) / 255
                    green = CGFloat((hexNumber & 0x00ff00) >> 8) / 255
                    blue = CGFloat((hexNumber & 0x0000ff) >> 0) / 255

                    self.init(red: red, green: green, blue: blue, alpha: 1.0)
                    return
                }
            }
        }

        return nil
    }
}
